tokio/src/runtime/task/join.rs

```diff
@@ -203,6 +203,38 @@ impl<T> JoinHandle<T> {
         }
     }
 
+    /// Checks if the task associated with the handle has finished(whether completed or cancelled).
+    /// This function does not block. Once this returns true, the task is expected
+    /// to finish quickly, without blocking for any significant amount of time.
+    ///
+    /// ```rust
+    /// use tokio::time;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///    let handle1 = tokio::spawn(async {
+    ///        // do some stuff here
+    ///    });
+    ///    let handle2 = tokio::spawn(async {
+    ///        // do some other stuff here
+    ///        time::sleep(time::Duration::from_secs(10)).await;
+    ///    });
+    ///    // Wait for the task to finish
+    ///    time::sleep(time::Duration::from_secs(1)).await;
+    ///    assert!(handle1.is_finished());
+    ///    handle2.abort();
+    ///    assert!(handle2.is_finished());
+    /// }
+    /// ```
+    pub fn is_finished(&self) -> bool {
+        if let Some(raw) = self.raw {
+            let state = raw.header().state.load();
+            state.is_complete() || state.is_cancelled()
+        } else {
```

Comment:
Hmm, I'm not so sure we want to use the `is_cancelled()` bit. The only situation where `is_cancelled()` is true but `is_complete()` is false is if `abort` has been called, but the task has not yet actually been cancelled.

Reply:
if `COMPLETED` is set when the task is cancelled, I think it's better to check `is_complete` only, as users can expect the handle is `Ready` when polled if `is_finished` returns true.